### ISSUE TITLE
Training a model to match CVPR paper action spotting average-mAP 

### ISSUE THREAD

**Issue by jvbsoares** [cioppaanthony/context-aware-loss#3]
Greetings,

Thank you for making this code and dataset available, it's fantastic!

I am trying to train a model to reproduce the average-mAP from the CVPR paper, but was not able to do so yet. I wanted to see if you might have any suggestions.

When I run testing (test.py) using the provided model (in weights/soccernet_trained_weights_resnet_pca512.h5) on the provided data, I am able to successfully match the paper's results. This gives me an average-mAP of 0.6289 on the provided test data, as expected.

However, when I train a new model with the default options using train.py (again, using the provided dataset), when I evaluate the model on the test set, I obtain around 0.60 average-mAP. The good news is that this result is pretty good, but still shy of the results in the paper.

I went over the paper and noticed that for the ResNet setup, lambda^{seg} is set to 1.5. If I'm not mistaken, in the code, this corresponds to "lossweightsegmentation", whose default value is 0.002. I tried training again while setting "lossweightsegmentation" to 1.5, but the result of that were just really low average-mAP.

Please let me know if you have any suggestions of what I might try out.

Thank you very much!
Joao

**Comment by cioppaanthony** [comment]
Hi Joao,

Thank you for your interest!  

First, let me clarify your question concerning the loss weight parameter. Actually it is just a difference in the definition of the loss. In the paper (See equation 9) there is a multiplying coefficient in front of the sum expression of 1/(C\*Nf) that is actually not in the code, as if you directly input lamda_seg/(C\*Nf) as a parameter. If you take the 1.5 and multiply it by this coefficient, you get back the 0.002 from the code (as 1.5/(3\*240)=0.0020833) . I should have made the explanations in the code clearer or even stick to the paper's description and do the computation of the coefficient in the code, I am sorry for that. I will try to clarify it in a future commit. So you do not have to change the value of lambda.

Concerning the performances, did you try to run it only one time ? Actually the performance that we show in the paper corresponds to an average performance since there is some variability between the trainings. If you run the training multiple times, the results should be comprised most of the time between 0.61 and 0.63. The code that we provide does not have a fixed seed for numpy and cuda, so each run may lead to different results. I am still a bit surprised by the 0.6, this is actually quite low and usually does not happen quite often. 

A last note on that is that the metric is quite sensitive to a single detection (correct or incorrect). For example when working on the paper we noticed one missed action in the labels of the testset. We informed the author of SoccerNet and he corrected the mistake (the testset is up to date on the SoccerNet's github and ours since then). this single error correction lead to a 1% increase in the performance, with the exact same weights. 

Therefore, if you want to reproduce the exact results, I suggest that you run multiple times the training and take the mean performance on the testset. Then you should get back the +-62,5%. 

This is the reason why we provided weights for a network that we trained that had performances close to the mean, so that people don't have to run it several times.

I hope that I answered your concerns. Don't hesitate to ask me if you have other questions or features that you would want in further releases.

Best regards,

Anthony

**Comment by jvbsoares** [comment]
Hi Anthony,

Thank you for the very quick reply. This information helps a lot!

Thank you for explaining about the weight loss parameter. That makes sense to me. I wasn't attentive enough to be able to see the connection between the code and paper, as you explained.

Indeed, I only ran training once. I should also note I was not in the docker environment suggested in the README, so it is possible that different library versions could be leading to some discrepancy. I will try multiple runs of training to see if that gives better results. Otherwise, I might try to work with docker as suggested in the README so that I have the same setup. In case it helps for future reference, my environment is on slightly more recent versions of the libraries:
tensorflow==2.2.0
numpy==1.19.1
opencv-python==4.4.0.42
tabulate==0.8.7
tqdm==4.48.2
h5py==2.10.0
matplotlib==3.3.1

Good point about the data. I checked the data I am using, and it looks to me like I was already on the latest version. (I downloaded the ResNET_PCA512.zip again today from https://drive.google.com/drive/folders/1VuVuwkJG1lTRNbwEno6fzcOe6zeBP45y?usp=sharing and verified that it matches what I have been working with.)

Thanks again for the help. It might take me some days to get back to this, but I will keep you posted.

Joao

**Comment by jvbsoares** [comment]
I wanted to give an update that I now ran the training multiple times (inside my slightly different environment), and the result is indeed much closer to the one in the paper now. Over 5 runs, I got an average of 0.6187, which seems close enough to the reported 0.625, given the variance over the different runs. Here are the individual scores I got:
[0.6301037280168093, 0.5949293752164433, 0.6238349457511257, 0.6308516919461646, 0.6137913952723644]

Thank you for your help. Let me know if any other questions or comments, but please feel free to close this issue.

**Comment by cioppaanthony** [comment]
Great, you're welcome! Happy to see that you that you are were able to reproduce the results.
The individual performances that you report seem great (except for the 0.59, but it can happen ;-) ).
